### ISSUE TITLE
Create an IP Checker Utility

### ISSUE THREAD

**Issue by rovinski** [The-OpenROAD-Project/OpenROAD#4872]
### Description

_Originally posted by @rovinski in https://github.com/The-OpenROAD-Project/OpenROAD/issues/4871#issuecomment-2026137319_

Currently, it is very possible to import an IP block which is poorly designed and will cause downstream issues in OpenROAD, which becomes difficult to root-cause. A checker utility which checks against a list of common problems can go a long way in identifying problems early.

### Suggested Solution

List of common problems to check (far from an exhaustive list):
- LEF
  - [ ] Macro dimensions are a multiple of the manufacturing grid
  - [ ] Pins are aligned to the manufacturing grid
  - [ ] Each signal pin must obey one of the following:
    - Minimum width, and each pin lies on the routing grid (with an offset is okay)
    - Greater than minimum size, has an access point on the routing grid, and connections from a default width wire OR an NDR rule does not cause a DRC violation.
    - Pins are wide enough such that regardless of offset, a connection from a default width wire (OR NDR rule) does not cause a DRC violation.
  - [ ] Power ring or grid straps are accessible (no obstructions in the way)
  - [ ] Signal pins are accessible (no obstructions in the way)
  - [ ] LEF file does not have an excessive number of polygons which will slow down OR
  - [ ] Ports have antenna info
  - [ ] LEFs in finfet processes have a fin grid property
- LIB
  - All quantities (rise, fall, delay, power, cap, etc.) are sane
    - [ ] No instantaneous transitions for reasonable input values
    - [ ] No excessively long transitions for reasonable input values
    - [ ] No excessively large capacitances
  - [ ] No missing properties
  - [ ] Characterization corner(s) match(es) the characterization corner(s) of the design
- LIB+LEF
  - [ ] Every LEF digital pin has lib timing info
  - [ ] LEF power pins defined match lib power pins
  - [ ] LEF and LIB area match

**Comment by QuantamHD** [comment]
Huge +1 from me.

**Comment by maliberty** [comment]
Thanks for writing this up.  A new comer doesn't have to implement every check in order to contribute.  Ticking any box is helpful.

**Comment by rovinski** [comment]
Indeed, I just wanted to list everything I could think of.

**Comment by razmikarshakyann** [comment]
Hi everyone,
I am currently working on this subtask: Each signal pin must obey one of the following:
- Minimum width, and each pin lies on the routing grid (with an offset is okay)
- Greater than minimum size, has an access point on the routing grid, and connections from a default width wire OR an NDR rule does not cause a DRC violation
- Pins are wide enough such that regardless of offset, a connection from a default width wire (OR NDR rule) does not cause a DRC violation

Here is what I have gathered during my research:
- As I understood minimum width is defined as MINWIDTH in LEF files.
- I found the checkPin function in io_parser_helper.cpp, which uses frPin to check pin alignment on the routing grid. However, I am aware of different pin types such dbMPin, dbBPin, etc.
- I discovered code in FlexPA_prep.cpp where prepPoint_pin is called and a variable nAps is used to determine if there are access points for a pin.

Despite the progress, I still have some critical questions that I need your expertise to clarify:
- Given that this task pertains to LEF files, which function(s) or method(s) should I use to gather all necessary pins before reading the DEF file? I could not find a specific function for this purpose and would appreciate any guidance
- Is the checkPin function the only method available to verify if a pin is located on the routing grid, or are there other recommended approaches? 
- Additionally, does the part of makeFastrouteNet function in the GlobalRouter serve any role in this validation?

@maliberty, @rovinski, @QuantamHD I appreciate your assistance in this matter. Any recommendations will greatly help me proceed with my task.

**Comment by maliberty** [comment]
@rovinski in order to talk about access points we would have to run pin_access.  This can only happen after macro placement at the earliest.  How are you intending for this checker to be used?  

Perhaps it would be easier to check if  the spacing between pins' centers are integral multiples of the track pitch.  Sufficiently wide pins could be allowed to be exceptions.

**Comment by titan73** [comment]
+1
We had to develop such a tool. Some checks are listed here and some others are for compatibility with proprietary tools.
And importantly, to check coherency between views (pins most of the time) between .lib, .lef, behavioural models, .cdl, ...
We have a internal standard for IP package structure but I don't know if there is an open standard for that though.

**Comment by rovinski** [comment]
Sorry for late reply.

Ideally pin rules could be checked in isolation without having to have the macro placed yet. It would, however, require tracks and NDRs to be created already (NDRs mainly to avoid false positives).

If the macro has to be placed in order to run pin access, maybe is it possible to place it in a virtual block which then gets deleted after pin access runs?

**Comment by maliberty** [comment]
If you are just checking that the pins could be placed on the track grid then you don't need to use pin access and can just look at the pin spacing and shapes.

**Comment by maliberty** [comment]
Pins are never blocked by obstructions - pin shapes have priority over obstructions.

**Comment by rovinski** [comment]
I mean something really dumb like this:

![image](https://github.com/user-attachments/assets/ba6f4313-27a6-4920-ad4f-b82bc0966068)

Also assuming that metals on top and bottom are blocked.

I've seen this happen in blocks that weren't extracted correctly.

**Comment by MarufHossain-tech** [comment]
I want to contribute to this issue. I can add as many checks as. I already created a demo. As soon as you assign it to me I will start finisher polish and do the work with in 3-5 days

**Comment by MarufHossain-tech** [comment]
### Summary
This PR adds a LEF Pin Checker utility script that performs:
- Pin-to-grid alignment checks
- Layer validation
- Direction rule enforcement

### Motivation
Supports design integrity at the LEF level. Built for robustness, clarity, and OpenROAD code style.

### Usage
```bash
python3 src/pin_checker/pin_checker.py

**Comment by MarufHossain-tech** [comment]
this is a draft . please give some feedbacks

**Comment by maliberty** [comment]
I'm confused as this is an issue not a PR.

**Comment by maliberty** [comment]
An additional check that would have been helpful today - the pin direction in LEF and Liberty match.

**Comment by MarufHossain-tech** [comment]
I am a newcomer to GitHub and the open-source contribution world. So I'm facing some newbee obstacles. My work is ready for PR, but I have a last query: do I need to make changes to the main source code, or can I create a separate code for IP check?  I assumed it to be a separate code and file. If it isn't, it will take me just a few hours to fix it

**Comment by maliberty** [comment]
I'm not sure what you mean by "a separate code and file".  Do you mean a new file within OpenROAD?  That would be fine.   Do you mean an entirely separate program?  That would seem to duplicate a lot of infrastructure.  It might be easier to just open a draft PR so I can see what you mean.

**Comment by mohamed-khairy-swe** [comment]
Hi,
Is this issue still valid?

**Comment by maliberty** [comment]
Yes as we don't have such a utility.

**Comment by mohamed-khairy-swe** [comment]
Ok I will see what I can do about it, but don't assign me because this will make me feel pressured 😅

**Comment by mohamed-khairy-swe** [comment]
@maliberty 
i'm currently working on the LEF checks and i was wondering if i should treat every check as an error or there are some checks that should be treated as a warning

**Comment by maliberty** [comment]
An error stops execution so you would only be able to see one problem at a time.  I suggest you mark these are warnings and at the end error if warnings > 0.

**Comment by mohamed-khairy-swe** [comment]
Hey @maliberty 👋,
I made changes to the PR last week but no one responded, so I was wondering if you're busy or Any change needs a new PR?

**Comment by rovinski** [comment]
There was some initial work in #9405 but still more to do.

**Comment by mohamed-khairy-swe** [comment]
I'm almost done with my degree, so once I've wrapped everything up, I'll return and continue improving the IP checker.

**Comment by mohamed-khairy-swe** [comment]
I'm currently too busy to continue working on this. If any contributor reading this, please feel free to continue from where I left off #9405 .

**Comment by kris70lesgo** [comment]
Hi @maliberty,
I noticed that PR #10820 is addressing the LEF/Liberty pin consistency checks. I'd like to contribute another independent IP checker from the checklist in this issue to avoid overlapping work.
Is there a particular remaining check that you'd recommend implementing next?
Thanks!

**Comment by maliberty** [comment]
@rovinski you were the original requestor - any priority?

**Comment by rovinski** [comment]
@kris70lesgo I updated the TODO list with the tasks that I believe have been completely resolved.
I think a main area you could contribute is that no one has started the Liberty sanity checking. A good place to start is checking for transition and capacitance values which are 0.

**Comment by rovinski** [comment]
There is also a lingering item from #9405 I described [here](https://github.com/The-OpenROAD-Project/OpenROAD/pull/9405#discussion_r2810759438) which is to check if multi-layer alignment is impossible.
